### PR TITLE
[arm64] Add tan intrinsic lowering

### LINKED ISSUE #94755: Add arm64 support for tan to cover f16 cases in a major isa

### PR DESCRIPTION
**Pull request by farzonl**
This change is an implementation of https://github.com/llvm/llvm-project/issues/87367's investigation on supporting IEEE math operations as intrinsics.
Which was discussed in this RFC:
https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

This PR is just for Tan.

Now that x86  tan backend landed: https://github.com/llvm/llvm-project/pull/90503 we can add other backends since the shared pieces are in tree now.

Changes:
- `llvm/include/llvm/Analysis/VecFuncs.def` - vectorization of tan for arm64 backends.
- `llvm/lib/Target/AArch64/AArch64FastISel.cpp` - Add tan to the libcall table
-  `llvm/lib/Target/AArch64/AArch64ISelLowering.cpp` - Add tan expansion for f128, f16, and vector\neon operations
-  `llvm/lib/Target/AArch64/GISel/AArch64LegalizerInfo.cpp` define `G_FTAN` as a legal arm64 instruction

resolves #94755

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1183]: Can you rejig this so there isn't a single opcode value on a line?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:735]: Is there a llvm.experimental.constrained.tan.f64 yet?

**Review by davemgreen — COMMENTED** [review]
I assume these are very similar to cos? It all looks sensible to me.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:735]: Thats in this PR: https://github.com/llvm/llvm-project/pull/94559

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1183]: That would cascading line changes, which I thought might make the change look bigger than it is. . That said I could but I was hoping to create some room for the other trig intrinsics I have planned to add from that RFC. No strong feeling though, so if youwant it i'll make it happen.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:735]: OK I see, great. And will that eventually involve adding testing for AArch64? Or is it best to add a quick STRICT_FTAN in here now? I know we have some tests for constrained.cos, and it doesn't look like much is needed to support them.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1183]: I think I would rejig them again later if needed. It's a shame clang-format does such an ugly job with it. It's nice to leave it clean in the meantime though.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:735]: Looks likere there are two places to add tests:
- llvm/test/CodeGen/AArch64/fp-intrinsics-fp16.ll
- llvm/test/CodeGen/AArch64/fp-intrinsics.ll

`fp-intrinsics.ll` I can add now and  to PR #94559 and everything will work.

fp-intrinsics-fp16.ll will need to wait for this  pr to merge before I add the tests because  f16 is handled on a per target basis.

**Review by paschalis-mpeis — CHANGES_REQUESTED** [review]
Can you please add the sleef tests as well, in the files below?
- replace-with-veclib-sleef-scalable.ll
- replace-with-veclib-sleef.ll

This should be the equivalent you did for ArmPL (replace-with-veclib-armpl.ll).


Regarding the TLI mappings, the symbols do exist in the libraries SLEEF and ArmPL (libsleefgnuabi.so and libamath.so respectively).

The code changes look reasonable to me as well, but no strong opinion on areas I haven't touched (fast/global isels, darwin changes).

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp
Comment (farzonl) [reply llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:1183]: done

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM if there are no other comments from @paschalis-mpeis on the tests.

**Review by paschalis-mpeis — APPROVED** [review]
@farzonl thanks a lot for adding the replace-with-veclib SLEEF tests. LGTM.

**Comment by farzonl** [comment]
@davemgreen  @paschalis-mpeis thank you both for the review!

**Pull request merged**
